Adding Configuration docs (#1181)
ansible/galaxy_ng: docs/config/collection_signing.md

# Enabling Collection Signing

> Galaxy Server can create, sync, receive as upload and verify and serve collection signatures.

Reviewer: ```suggestion
Galaxy Server can create, sync, receive as well as upload, verify, and serve collection signatures.
```
Maybe this?

A signature is typically a GPG ASCII Armoured Detached artifact, in other words, a file
e.g: `MANIFEST.json.asc` that is created by a GPG compatible script based on the `MANIFEST.json`
file.

The `MANIFEST.json` file already contains checksums to verify all the other artifacts inside a 
collection so that is why the signature is created based on this file. 

!!! tip "Enabling signing on pulp installer"
    Pulp installer can also be configured to enable signing https://github.com/pulp/pulp_installer/tree/main/roles/galaxy_post_install#variables-for-the-signing-service
    if you configures using pulp-installer you can skip the `Creating Signing Service` section of this page.

## Creating Signatures

Signatures can be manually created and then later uploaded to a galaxy server or the galaxy
server can internally create a signature for each collection.

### Signing Service

A signing service is an object defined on the **pulp** backend that combines a GPG **key** and the

> absolute path an executable script.

Reviewer: ```suggestion
absolute path of an executable script.
```

#### Creating a signing service

To create a signing service on the Galaxy server it is needed to access the `django-admin` utility
and execute:

!!! info "important"
    The command must run inside the Python Environment where galaxy_ng workers are installed.

```bash title="Bash Command"
django-admin add-signing-service \  #(1)
    unique-name \  #(2)
    /abs/path/to/script.sh \ #(3)
    GPG_KEY_ID #(4)
```

1. A django management command.

> 2. Name, the signing_service should get in the database.

Reviewer: ```suggestion
2. Name the signing_service should get in the database.
```

> 3. Absolute path to Shell script where the signing is created.

Reviewer: Maybe either "where the signature is created" or "where the signing is performed"

> 3. Absolute path to Shell script where the signing is created.

Author: ```suggestion
3. Absolute path to Shell script where the signing is performed.
```

4. Key id of the public key.

The positional arguments to the `add-signing-service` command:

> - **name**: Must a unique name to identify the signing service object.

Reviewer: ```suggestion
- **name**: Must be a unique name to identify the signing service object.
```

- **script**: Must be a path to an executable acessible and runnable by the pulp worker process user.
    One example script can be found at the pulp documentation: [https://docs.pulpproject.org/pulpcore/workflows/signed-metadata.html#metadata-signing](https://docs.pulpproject.org/pulpcore/workflows/signed-metadata.html#metadata-signing)
- **GPG_KEY_ID**: Must be the ID or e-mail that identifies the key on the global user keyring.

>     The GPG key must be located at the pulp user level keyring and must be a valid GPG there is

Reviewer: ```suggestion
    The GPG key must be located at the pulp user level keyring and must be a valid GPG, there is
```

    a guide on how to create a valid gpg key here [https://access.redhat.com/articles/3359321](https://access.redhat.com/articles/3359321)

!!! important
    When importing the key to the keyring the trust level must be set to some value higher than 3.  
    ex: `echo "${KEY_FINGERPRINT}:6:" | gpg --batch --import-ownertrust`

#### Configuring Galaxy to use the Signing Service

To tell galaxy which signing service to use you need to set it in the pulp settings.

Option 1:
```py title="/etc/pulp/settings.py"
GALAXY_COLLECTION_SIGNING_SERVICE = "unique-name"  #(1)
```

1. The name of the created signing service

Option 2:
```bash title="Environment Variable"
export PULP_GALAXY_COLLECTION_SIGNING_SERVICE=unique-name #(1)
```

1. The name of the created signing service

> #### Configuring Galaxy to automatic sign during approval

Reviewer: ```suggestion
#### Configuring Galaxy to automatically sign during approval
```

    STATUS=$?
    if [ $STATUS -eq 0 ]; then
    echo {\"file\": \"$FILE_PATH\", \"signature\": \"$SIGNATURE_PATH\"}
    else
    exit $STATUS
    fi
    EOF

    ##########################################
    # Create the signing service
    ##########################################

    django-admin add-signing-service \
    ansible-default \
    /etc/pulp/scripts/collection_sign.sh \
    $KEY_FINGERPRINT

    ################################################
    # Enable the signing service on galaxy settings
    ################################################

    echo "GALAXY_COLLECTION_SIGNING_SERVICE='ansible-default'" >> /etc/pulp/settings.py

    # OR

    export PULP_GALAXY_COLLECTION_SIGNING_SERVICE=ansible-default

    ####################################################
    # Optionally enable automatic signing upon approval
    # NOTE: this must be enabled only if 
    # `GALAXY_REQUIRE_CONTENT_APPROVAL` is True.
    ####################################################

    echo "GALAXY_AUTO_SIGN_COLLECTIONS=True" >> /etc/pulp/settings.py

    # OR

    export PULP_GALAXY_AUTO_SIGN_COLLECTIONS=True
    ```

#### Signing via API

```bash
curl -X POST \

>   'http://0.0.0.0:8002/api/automation-hub/_ui/v1/collection_signing/' \  # (1)

Author: this one is for upload, the other is for signing

    # (the same private key that matched the repo keyring previously configured)
    gpg --quiet --batch --pinentry-mode loopback --yes 
    --homedir ~/.gnupg/ --detach-sign --default-key KEY_ID \
    --armor --output MANIFEST.json.asc MANIFEST.json
    ```
4. The signing generates a file `MANIFEST.json.asc` that now can be uploaded to Galaxy Server.

#### Uploading the signature via API

```bash
curl -X 'POST' \
  'http://0.0.0.0:5001/pulp/api/v3/content/ansible/collection_signatures/' \
  -H 'accept: application/json' \
  -H 'Authorization: Basic YWRtaW46YWRtaW4=' \
  -H 'Content-Type: multipart/form-data' \
  -F 'file=@MANIFEST.json.asc;type=text/plain' \
  -F 'repository=/pulp/api/v3/repositories/ansible/ansible/43dd716c-f7dd-4a80-b889-36818ed3e347/' \
  -F 'signed_collection=/pulp/api/v3/content/ansible/collection_versions/40f85531-29a5-4eb7-b4cd-a850f8824daf/'
```

!!! warning
    This endpoint is under `/pulp/` namespace and requires the relations to be set
    using `pulp_hrefs` (full url for pulp objects)

#### Uploading the signature via UI

When enabled there will be upload buttons on the pages:

- Collection Version
- Approval Dashboard

## Syncing Signatures

> When syncing from remote servers, if the signature is server Galaxy will also sync all the 

Reviewer: not sure what this means "if the signature is server"

> When syncing from remote servers, if the signature is server Galaxy will also sync all the 

Author: ```suggestion
When syncing from remote servers, if the signature is served by the remote then Galaxy will also sync all the 
```
